keps/sig-api-machinery/storage-migration-auto-trigger.md

```diff
@@ -0,0 +1,148 @@
+}
+
+type StorageStateStatus {
+  // The hash values of versions that the persisted objects might be encoded in.
+  // If empty, it means the stored versions are unknown.
+  // +optional
+  StoredVersionsHashes []string
+  // The hash value of the storage version, as shown in the discovery document
+  // served by the API server.
+  // Storage Version is the version to which objects are converted to
+  // before persisted.
+  StorageVersionHash string
+}
+```
+
+We had considered making `StoredVersionsHashes` part of the `spec` instead of
+the `status`, because the stored versions cannot be deduced immediately by
+inspecting Kubernetes API. However, we make it part of the status because
+* Its value can be determined by inspecting the `StorageVersionMigration` API.
+  It just takes longer, i.e., until a migration has completed. 
+* Leaving it in the status is consistent with the [CRD.status.StoredVersions][]
+  API.
+
+[CRD.status.StoredVersions]:https://github.com/kubernetes/kubernetes/blob/697c2316faaabae8ef8371032b60be65d7795e68/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L305
+  
+## Storage migration triggering controller
+
+We will add a controller that monitors the discovery documents and the
+storageStates to trigger storage version migration.
+
+The controller fetches the discovery document periodically. For each resource,
+it compares the storageVersion in the discovery document with that recorded in
+the storageState. If they differ, or if the storageState does not exist, the
+controller
+  * stops any ongoing migration of this resource by deleting existing
+    storageVersionMigrations,
+  * creates a new storageVersionMigration for this resource,
+  * in a single operation, updates (or creates) the
+    storageState.status.StorageVersionHash to the value shown in the discovery
+    document, and adds the hash to the storageState.status.storedVersionHashes.
+
+The refresh rate of the discovery document needs to satisfy two conflicting
+goals: *i)* not causing too much traffic to apiservers, and *ii)* not missing
+storage version configuration changes. More precisely, the storage version of a
+certain resource should not change twice within a refresh period. 10 minutes is
+a good middle ground.
+
+The controller uses an informer to monitor all storageVersionMigrations. If a
+migration completes, the controller finds the storageState of that resource,
+updates its status.storedVersionHashes to equal its status.storageVersionHash.
+
+If the controller crashloops and misses storage version changes, the
+storageState.status.storedVersionHashes can be stale. Thus, the controller
+should recreate storageState objects with empty storedVersionHashes when it
+bootstraps to avoid acting on stale information.
+
+## Implications to cluster operators
+
+* cluster operators need to make sure consecutive upgrade/downgrade attempts are
+  at least 10 minutes apart. Otherwise because the migration triggering
```

Comment:
Is this a reasonable requirement? If I push the wrong thing I'm going to immediately stop it and push the right thing, I don't want to run the wrong thing for 10 minutes.